pkg/providers/subnet/subnet.go

```diff
@@ -103,29 +111,48 @@ func (p *DefaultProvider) List(ctx context.Context, nodeClass *v1beta1.EC2NodeCl
 
 // CheckAnyPublicIPAssociations returns a bool indicating whether all referenced subnets assign public IPv4 addresses to EC2 instances created therein
 func (p *DefaultProvider) CheckAnyPublicIPAssociations(ctx context.Context, nodeClass *v1beta1.EC2NodeClass) (bool, error) {
-	subnets, err := p.List(ctx, nodeClass)
-	if err != nil {
-		return false, err
+	subnets := map[string]bool{}
+	for _, subnet := range nodeClass.Status.Subnets {
+		if subnetAssociatePublicIP, ok := p.associatePublicIPAddressCache.Get(subnet.ID); ok {
+			subnets[subnet.ID] = subnetAssociatePublicIP.(bool)
+		}
 	}
-	_, ok := lo.Find(subnets, func(s *ec2.Subnet) bool {
-		return aws.BoolValue(s.MapPublicIpOnLaunch)
-	})
-	return ok, nil
+
+	for _, v := range subnets {
+		if v {
+			return true, nil
+		}
+	}
+	return false, nil
 }
 
 // ZonalSubnetsForLaunch returns a mapping of zone to the subnet with the most available IP addresses and deducts the passed ips from the available count
 func (p *DefaultProvider) ZonalSubnetsForLaunch(ctx context.Context, nodeClass *v1beta1.EC2NodeClass, instanceTypes []*cloudprovider.InstanceType, capacityType string) (map[string]*ec2.Subnet, error) {
-	subnets, err := p.List(ctx, nodeClass)
-	if err != nil {
-		return nil, err
-	}
-	if len(subnets) == 0 {
-		return nil, fmt.Errorf("no subnets matched selector %v", nodeClass.Spec.SubnetSelectorTerms)
-	}
 	p.Lock()
 	defer p.Unlock()
-	// sort subnets in ascending order of available IP addresses and populate map with most available subnet per AZ
+
+	if len(nodeClass.Status.Subnets) == 0 {
+		return nil, fmt.Errorf("no subnets matched selector %v", nodeClass.Spec.SubnetSelectorTerms)
+	}
+
 	zonalSubnets := map[string]*ec2.Subnet{}
+
+	availableIPAddress := map[string]int64{}
+	for _, subnet := range nodeClass.Status.Subnets {
+		if subnetAvailableIP, ok := p.availableIPAddressCache.Get(subnet.ID); ok {
+			availableIPAddress[subnet.ID] = subnetAvailableIP.(int64)
+		}
+	}
+
+	subnets := lo.Map(nodeClass.Status.Subnets, func(s v1beta1.Subnet, _ int) *ec2.Subnet {
```

Comment:
Is it worth creating a tightly scoped struct here to store only the details we need rather than using an ec2.Subnet? The only concern that I have here is that someone not really reading the code could think that this has all of the data in the ec2.Subnet, when it reality we are only injecting the data that we care about